### PR TITLE
Support LATERAL join (v3)

### PR DESCRIPTION
**Pull request by kokosing**

### CODE REVIEW

**Comment by kokosing** [comment]
Supersedes https://github.com/prestodb/presto/pull/6841

**Comment by kokosing** [comment]
Since this is a long pull request, I divided it into two smaller ones. See #6921

**Comment by kokosing** [comment]
rebased

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 1]
```diff
@@ -357,7 +357,47 @@ else if (firstDependencies.stream().allMatch(right::canResolve) && secondDepende
...
+        PlanBuilder leftPlanBuilder = initializePlanBuilder(leftPlan);
+        PlanBuilder rightPlanBuilder = initializePlanBuilder(rightPlan);
+
+        PlanBuilder planBuilder = subqueryPlanner.appendLateralJoin(leftPlanBuilder, rightPlanBuilder, lateral.getQuery(), true);
+
+        List<Symbol> outputSymbols = ImmutableList.<Symbol>builder()
+                .addAll(leftPlan.getOutputSymbols())
+                .addAll(rightPlan.getOutputSymbols())
+                .build();
+        return new RelationPlan(
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: Arguments can all go on the same line.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 2]
```diff
@@ -401,16 +441,6 @@ else if (type instanceof MapType) {
         return new RelationPlan(unnestNode, analysis.getScope(joinNode), unnestNode.getOutputSymbols());
     }
 
-    private static Expression oneIfNull(Optional<Symbol> symbol)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: Unrelated change?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedSingleRowSubqueryToProject.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+ *     - Project (A + C)
+ *       - single row VALUES()
+ * </pre>
+ * to:
+ * <pre>
+ *   - Project(A, B, C, A + C)
+ *       - (input) plan which produces symbols: [A, B, C]
+ * </pre>
+ */
+public class TransformCorrelatedSingleRowSubqueryToProject
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedSingleRowSubqueryToProject.java]: I know this predates the iterative optimizer, but can we port it to that framework?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedApplyToJoin.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class TransformUncorrelatedApplyToJoin
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedApplyToJoin.java]: I know this predates the iterative optimizer, but can we port it to that framework?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java [hunk 5]
```diff
@@ -96,15 +96,6 @@ private static boolean isSupportedSubqueryExpression(Expression expression)
                 expression instanceof QuantifiedComparisonExpression;
     }
 
-    /**
-     * @return true when subquery is scalar and it's output symbols are directly mapped to ApplyNode output symbols
-     */
-    public boolean isResolvedScalarSubquery()
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: Unrelated change?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedNoAggregationSubqueryToJoin.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+ * <pre>
+ *   - Join(INNER, B = C)
+ *       - (input) plan which produces symbols: [A, B]
+ *       - Filter(D < 5)
+ *          - plan which produces symbols: [C, D]
+ * </pre>
+ * <p>
+ * Note only conjunction predicates in FilterNode are supported
+ */
+public class TransformCorrelatedNoAggregationSubqueryToJoin
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedNoAggregationSubqueryToJoin.java]: I know this predates the iterative optimizer, but can we port it to that framework?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedSingleRowSubqueryToProject.java (hunk 3 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedSingleRowSubqueryToProject.java]: Sure thing, but I would prefer to merge this in as it is and the port it. Since we have a switch to run only legacy rules we need still both versions anyway.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedApplyToJoin.java (hunk 4 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedApplyToJoin.java]: Sure thing, but I would prefer to merge this in as it is and the port it. Since we have a switch to run only legacy rules we need still both versions anyway.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedNoAggregationSubqueryToJoin.java (hunk 6 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedNoAggregationSubqueryToJoin.java]: Sure thing, but I would prefer to merge this in as it is and the port it. Since we have a switch to run only legacy rules we need still both versions anyway.

**Comment by kokosing** [comment]
Superseded by #8218. I superseded this PR because the rebase was quite massive and I did not want to lose the original code (before the rebase).
